pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -283,39 +283,65 @@ func balanceDomains(
-				nodesPodFitsOnBesidesCurrent(aboveToEvict[k], nodeMap) == 0 {
-				klog.V(2).InfoS("Ignoring pod for eviction due to node selector/affinity", "pod", klog.KObj(aboveToEvict[k]))
+			if err := validatePodFitsOnOtherNodes(aboveToEvict[k], nodeMap); err != nil {
+				klog.V(2).InfoS(err.Error(), "pod", klog.KObj(aboveToEvict[k]))
 				continue
 			}
+
 			podsForEviction[aboveToEvict[k]] = struct{}{}
 		}
 		sortedDomains[j].pods = sortedDomains[j].pods[:len(sortedDomains[j].pods)-movePods]
 		sortedDomains[i].pods = append(sortedDomains[i].pods, aboveToEvict...)
 	}
 }
 
-// nodesPodFitsOnBesidesCurrent counts the number of nodes this pod could fit on based on its affinity
+// validatePodFitsOnOtherNodes performs validation based on scheduling predicates for affinity and toleration.
 // It excludes the current node because, for the sake of domain balancing only, we care about if there is any other
 // place it could theoretically fit.
 // If the pod doesn't fit on its current node, that is a job for RemovePodsViolatingNodeAffinity, and irrelevant to Topology Spreading
-func nodesPodFitsOnBesidesCurrent(pod *v1.Pod, nodeMap map[string]*v1.Node) int {
-	count := 0
+func validatePodFitsOnOtherNodes(pod *v1.Pod, nodeMap map[string]*v1.Node) error {
+	// if the pod has a hard nodeAffinity/nodeSelector/toleration that only matches this node,
+	// don't bother evicting it as it will just end up back on the same node
+	// however we still account for it "being evicted" so the algorithm can complete
+	// TODO(@damemi): Since we don't order pods wrt their affinities, we should refactor this to skip the current pod
+	// but still try to get the required # of movePods (instead of just chopping that value off the slice above)
+	isRequiredDuringSchedulingIgnoredDuringExecution := pod.Spec.Affinity != nil &&
+		pod.Spec.Affinity.NodeAffinity != nil &&
+		pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil
+
+	hardTaintsFilter := func(taint *v1.Taint) bool {
+		return taint.Effect == v1.TaintEffectNoSchedule || taint.Effect == v1.TaintEffectNoExecute
+	}
+
+	var availableTargetNodes, ineligibleAffinityCount, taintNotToleratedCount int
 	for _, node := range nodeMap {
-		if nodeutil.PodFitsCurrentNode(pod, node) && node != nodeMap[pod.Spec.NodeName] {
-			count++
+		if node == nodeMap[pod.Spec.NodeName] {
+			continue
+		}
+
+		if pod.Spec.NodeSelector != nil || isRequiredDuringSchedulingIgnoredDuringExecution {
+			if !nodeutil.PodFitsCurrentNode(pod, node) {
+				ineligibleAffinityCount++
+				continue
+			}
+		}
+
+		if !utils.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, hardTaintsFilter) {
+			taintNotToleratedCount++
+			continue
 		}
+
+		availableTargetNodes++
+	}
+
+	if availableTargetNodes == 0 {
+		return fmt.Errorf(
+			"ignoring pod for eviction due to %d nodes with ineligible selector/affinity and %d nodes with taints that are not tolerated",
```

Comment:
you could set the error specifically based on the above checks, so we know if it was either skipped due to affinity or taints. Perhaps even aggregate them in case it's both: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/errors#NewAggregate

Reply:
note that it could be because of the combination of both:

let's say we have 6 target nodes:
3 nodes with ineligible nodeSelector/Affinity
3 nodes with taints

=> not evicted

Reply:
Right, but it's an odd error to return, for example: `ignoring due to 0 nodes with selector and 6 nodes with affinity`

It would be more descriptive to list which nodes are failing for what, and don't include any unnecessary information. Using an aggretate error struct would be the easiest way to do that I think